go/engine/per_user_key_background.go

```diff
@@ -0,0 +1,193 @@
+}
+
+// GetPrereqs returns the engine prereqs.
+func (e *PerUserKeyBackground) Prereqs() Prereqs {
+	return Prereqs{}
+}
+
+// RequiredUIs returns the required UIs.
+func (e *PerUserKeyBackground) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{
+		libkb.LogUIKind,
+	}
+}
+
+// SubConsumers returns the other UI consumers for this engine.
+func (e *PerUserKeyBackground) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{&PaperKeyGen{}}
+}
+
+// Run starts the engine.
+// Returns immediately, kicks off a background goroutine.
+func (e *PerUserKeyBackground) Run(ectx *Context) (err error) {
+	ctx := ectx.NetContext
+	defer e.G().CTrace(ctx, "PerUserKeyBackground", func() error { return err })()
+
+	// use a new background context with a saved cancel function
+	ctx, cancel := context.WithCancel(context.Background())
+	e.Lock()
+	e.shutdownFunc = cancel
+	if e.shutdown {
+		// Shutdown before started
+		cancel()
+		e.Unlock()
+		e.meta("early-shutdown")
+		return nil
+	}
+	e.Unlock()
+
+	// start the loop and return
+	go func() {
+		e.meta("loop-start")
+		err := e.loop(ctx, ectx)
+		if err != nil {
+			e.G().Log.CDebugf(ctx, "PerUserKeyBackground loop error: %s", err)
+		}
+		cancel()
+		e.meta("loop-exit")
+	}()
+
+	return nil
+}
+
+func (e *PerUserKeyBackground) Shutdown() {
+	e.Lock()
+	defer e.Unlock()
+	e.shutdown = true
+	if e.shutdownFunc != nil {
+		e.shutdownFunc()
+	}
+}
+
+func (e *PerUserKeyBackground) loop(ctx context.Context, ectx *Context) error {
+	if err := libkb.SleepWithContext(ctx, e.G().Clock(), e.args.Settings.Start); err != nil {
+		return err
+	}
+	e.meta("woke-start")
+	var i int
+	for {
+		i++
+		err := e.round(ctx, ectx)
+		if err != nil {
+			e.G().Log.CDebugf(ctx, "PerUserKeyBackground round(%v) error: %s", i, err)
+		} else {
+			e.G().Log.CDebugf(ctx, "PerUserKeyBackground round(%v) complete", i, err)
+		}
+		if e.args.testingRoundResCh != nil {
+			e.args.testingRoundResCh <- err
+		}
+		if err := libkb.SleepWithContext(ctx, e.G().Clock(), e.args.Settings.Interval); err != nil {
+			return err
+		}
+		e.meta("woke-interval")
+		if err := libkb.SleepWithContext(ctx, e.G().Clock(), e.args.Settings.WakeUp); err != nil {
+			return err
+		}
+		e.meta("woke-wakeup")
+	}
+}
+
+func (e *PerUserKeyBackground) round(ctx context.Context, ectx *Context) error {
+	ctx, cancel := context.WithTimeout(ctx, e.args.Settings.Limit)
+	defer cancel()
+
+	if !e.G().Env.GetUpgradePerUserKey() {
+		e.G().Log.CDebugf(ctx, "CheckUpgradePerUserKey disabled")
+		return nil
+	}
+
+	if e.G().ConnectivityMonitor.IsConnected(ctx) == libkb.ConnectivityMonitorNo {
+		e.G().Log.CDebugf(ctx, "CheckUpgradePerUserKey giving up offline")
+		return nil
+	}
+
+	// Do a fast local check to see if our work is done.
+	pukring, err := e.G().GetPerUserKeyring()
+	if err == nil {
+		if pukring.HasAnyKeys() {
+			e.G().Log.CDebugf(ctx, "CheckUpgradePerUserKey already has keys")
+			return nil
+		}
+	}
+
+	arg := &PerUserKeyUpgradeArgs{}
```

Comment:
Through manual testing I found that `LogUI` doesn't exist so the engine bails. So I just took it out of the required list and logging still seems to work. 👍